README

fswebcam - Small and simple webcam software for *nix.

Created by Philip Heron <phil@sanslogic.co.uk>
http://www.sanslogic.co.uk/fswebcam/

This is the program used to generate images for a webcam. It captures a number
of frames from any V4L or V4L2 compatible device, averages them to reduce noise
and draws the details on it using the GD Graphics Library which also handles
compressing the image to PNG or JPEG.

INSTALLING

Run the following commands in the source folder to build and install fswebcam:

./configure --prefix=/usr --disable-v4l1 --enable-32bit-buffer
make
sudo make install

It's only requirements are that the GD library be installed with JPEG, PNG
and FreeType support.

Install Dependencies:
====================
sudo apt-get install libgd2-xpm-dev libjpeg-dev libpng-dev 

sudo apt-get install libfreetype6-dev

Run samples:
===========
fswebcam --Vflip 1 -r 1600x1200 -p YUV420P - > cam640x480_2.jpg
fswebcam --Hflip 1 -r 640x480 -p YUV420P - > cam640x480_1.jpg
fswebcam --exposure 4 -r 640x480 -p YUV420P - > cam640x480_3.jpg

** Don't forget to update your GC2035 driver to get better results


## Mainline Kernel support

In order to use fswebcam in your Distro running a mainline kernel version you need to play with **media-ctl** and **v4l2-ctl** before you grab the files.

* Install v4l2 utils

        sudo apt-get install v4l-utils
        
* Find your camera sensor

        v4l2-ctl --list-devices
        cedrus (platform:cedrus):
          /dev/video1

        sun6i-csi (platform:csi):
          /dev/video0

* List the formats available

        v4l2-ctl -d 0 --list-formats --list-ctrls

        User Controls

                               contrast 0x00980901 (int)    : min=0 max=255 step=1 default=0 value=0 flags=slider
                             saturation 0x00980902 (int)    : min=0 max=255 step=1 default=64 value=64 flags=slider
                                    hue 0x00980903 (int)    : min=0 max=359 step=1 default=0 value=0 flags=slider
                white_balance_automatic 0x0098090c (bool)   : default=1 value=1 flags=update
                            red_balance 0x0098090e (int)    : min=0 max=4095 step=1 default=0 value=0 flags=inactive, slider
                           blue_balance 0x0098090f (int)    : min=0 max=4095 step=1 default=0 value=0 flags=inactive, slider
                               exposure 0x00980911 (int)    : min=0 max=65535 step=1 default=0 value=590 flags=inactive, volatile
                         gain_automatic 0x00980912 (bool)   : default=1 value=1 flags=update
                                   gain 0x00980913 (int)    : min=0 max=1023 step=1 default=0 value=23 flags=inactive, volatile
                        horizontal_flip 0x00980914 (bool)   : default=0 value=0
                          vertical_flip 0x00980915 (bool)   : default=0 value=0
                   power_line_frequency 0x00980918 (menu)   : min=0 max=3 default=1 value=1

        Camera Controls

                          auto_exposure 0x009a0901 (menu)   : min=0 max=1 default=0 value=0 flags=update

        Image Processing Controls

                             pixel_rate 0x009f0902 (int64)  : min=0 max=2147483647 step=1 default=61430400 value=61430400 flags=read-only
                           test_pattern 0x009f0903 (menu)   : min=0 max=4 default=0 value=0
        ioctl: VIDIOC_ENUM_FMT
          Index       : 0
          Type        : Video Capture
          Pixel Format: 'BA81'
          Name        : 8-bit Bayer BGBG/GRGR

          Index       : 1
          Type        : Video Capture
          Pixel Format: 'GBRG'
          Name        : 8-bit Bayer GBGB/RGRG

          Index       : 2
          Type        : Video Capture
          Pixel Format: 'GRBG'
          Name        : 8-bit Bayer GRGR/BGBG

          Index       : 3
          Type        : Video Capture
          Pixel Format: 'RGGB'
          Name        : 8-bit Bayer RGRG/GBGB

          Index       : 4
          Type        : Video Capture
          Pixel Format: 'BG10'
          Name        : 10-bit Bayer BGBG/GRGR

          Index       : 5
          Type        : Video Capture
          Pixel Format: 'GB10'
          Name        : 10-bit Bayer GBGB/RGRG

          Index       : 6
          Type        : Video Capture
          Pixel Format: 'BA10'
          Name        : 10-bit Bayer GRGR/BGBG

          Index       : 7
          Type        : Video Capture
          Pixel Format: 'RG10'
          Name        : 10-bit Bayer RGRG/GBGB

          Index       : 8
          Type        : Video Capture
          Pixel Format: 'BG12'
          Name        : 12-bit Bayer BGBG/GRGR

          Index       : 9
          Type        : Video Capture
          Pixel Format: 'GB12'
          Name        : 12-bit Bayer GBGB/RGRG

          Index       : 10
          Type        : Video Capture
          Pixel Format: 'BA12'
          Name        : 12-bit Bayer GRGR/BGBG

          Index       : 11
          Type        : Video Capture
          Pixel Format: 'RG12'
          Name        : 12-bit Bayer RGRG/GBGB

          Index       : 12
          Type        : Video Capture
          Pixel Format: 'YUYV'
          Name        : YUYV 4:2:2

          Index       : 13
          Type        : Video Capture
          Pixel Format: 'YVYU'
          Name        : YVYU 4:2:2

          Index       : 14
          Type        : Video Capture
          Pixel Format: 'UYVY'
          Name        : UYVY 4:2:2

          Index       : 15
          Type        : Video Capture
          Pixel Format: 'VYUY'
          Name        : VYUY 4:2:2

          Index       : 16
          Type        : Video Capture
          Pixel Format: 'HM12'
          Name        : YUV 4:2:0 (16x16 Macroblocks)

          Index       : 17
          Type        : Video Capture
          Pixel Format: 'NV12'
          Name        : Y/CbCr 4:2:0

          Index       : 18
          Type        : Video Capture
          Pixel Format: 'NV21'
          Name        : Y/CrCb 4:2:0

          Index       : 19
          Type        : Video Capture
          Pixel Format: 'YU12'
          Name        : Planar YUV 4:2:0

          Index       : 20
          Type        : Video Capture
          Pixel Format: 'YV12'
          Name        : Planar YVU 4:2:0

          Index       : 21
          Type        : Video Capture
          Pixel Format: 'NV16'
          Name        : Y/CbCr 4:2:2

          Index       : 22
          Type        : Video Capture
          Pixel Format: 'NV61'
          Name        : Y/CrCb 4:2:2

          Index       : 23
          Type        : Video Capture
          Pixel Format: '422P'
          Name        : Planar YUV 4:2:2

          Index       : 24
          Type        : Video Capture
          Pixel Format: 'RGBP'
          Name        : 16-bit RGB 5-6-5

          Index       : 25
          Type        : Video Capture
          Pixel Format: 'RGBR'
          Name        : 16-bit RGB 5-6-5 BE

          Index       : 26
          Type        : Video Capture
          Pixel Format: 'JPEG' (compressed)
          Name        : JFIF JPEG
          
* find the info and topology

          media-ctl --device /dev/media1 -p
          Media controller API version 5.7.9

          Media device information
          ------------------------
          driver          sun6i-csi
          model           Allwinner Video Capture Device
          serial          
          bus info        
          hw revision     0x0
          driver version  5.7.9

          Device topology
          - entity 1: sun6i-csi (1 pad, 1 link)
                      type Node subtype V4L flags 0
                      device node name /dev/video0
            pad0: Sink
              <- "ov5640 2-003c":0 [ENABLED,IMMUTABLE]

          - entity 5: ov5640 2-003c (1 pad, 1 link)
                      type V4L2 subdev subtype Sensor flags 0
                      device node name /dev/v4l-subdev0
            pad0: Source
              [fmt:UYVY8_2X8/640x480@1/30 field:none colorspace:srgb xfer:srgb ycbcr:601 quantization:full-range]
              -> "sun6i-csi":0 [ENABLED,IMMUTABLE]


* Grab files from **fswebcam**

          media-ctl --device /dev/media1 --set-v4l2 '"ov5640 2-003c":0[fmt:YUYV8_2X8/1280x720]'
          ./fswebcam --displayfps 1 -S 30 -d /dev/video0 -r 1280x720 -p YUV420P - > /tmp/cam_640x480_yuv420p.jpg



debian/changelog
fswebcam (20140113-2) unstable; urgency=medium

  * Build-Depend on libgd-dev instead of libgd2-*-dev (Closes: #881745)
  * Use DEP5 format for debian/copyright
  * Bump Standards-Version (no changes needed)

 -- Luca Niccoli <lultimouomo@gmail.com>  Wed, 22 Nov 2017 10:56:11 +0100

fswebcam (20140113-1) unstable; urgency=low

  * Update debian/watch
  * Import Upstream version 20140113
  * Bump Standards-Version to 3.9.5 (no changes needed)
  * Move to debhelper 9
  * Import CPPFLAGS during build
  * Move packaging to git

 -- Luca Niccoli <lultimouomo@gmail.com>  Tue, 21 Jan 2014 00:02:55 +0100

fswebcam (20110717-1) unstable; urgency=low

  * New upstream release (Closes: #655138)
  * Bump Standards-Version to 3.9.3 (no changes needed).

 -- Luca Niccoli <lultimouomo@gmail.com>  Sun, 22 Apr 2012 23:59:34 +0200

fswebcam (20101118-2) unstable; urgency=low

  * videodev.h has moved out of build-essentials, get it from libv4l-dev.
    Thanks to Daniel Polehn that fixed the problem in Ubuntu and forwarded
    the patch to Debian. (Closes: #631577)
  * Bump Standards-Version to 3.9.2 (no changes needed).

 -- Luca Niccoli <lultimouomo@gmail.com>  Sat, 25 Jun 2011 19:13:25 +0200

fswebcam (20101118-1) unstable; urgency=low

  * New upstream release (Closes: #604121)
  * Enable libgd fontconfig support if available. Default to "sans:10".
    You'll need libgd2-xpm to use this feature.
  * Add S561 compressed palette support.
  * Fix small memory leak in the logging system.
  * Add option to swap colour channels in the image.
  * Bump Standards-Version to 3.9.1 (make use of architecture wildcard).

 -- Luca Niccoli <lultimouomo@gmail.com>  Sun, 06 Feb 2011 15:06:25 +0100

fswebcam (20100622-1) unstable; urgency=low

  * New upstream release. (Closes: #581518)
  * Remove patches merged upstream.
  * Fixes to the V4L1 source. (Closes: #586366)

 -- Luca Niccoli <lultimouomo@gmail.com>  Wed, 23 Jun 2010 23:00:13 +0200

fswebcam (20100401-1) unstable; urgency=low

  * New upstream release.
  * Remove patches merged upstream.
  * Update upstream email address.
  * Bump Standards-Version to 3.8.4 (no changes needed).

 -- Luca Niccoli <lultimouomo@gmail.com>  Mon, 05 Apr 2010 13:18:29 +0200

fswebcam (20091224-1) unstable; urgency=low

  * New upstream release. (Closes: #564889)
  * Switch to dpkg-source 3.0 (quilt) format.
  * Switch to dh7 minimal rules file.
  * Don't draw an empty banner.
  * Bump Standards-Version to 3.8.3 (no changes needed).

 -- Luca Niccoli <lultimouomo@gmail.com>  Wed, 13 Jan 2010 22:24:14 +0100

fswebcam (20070108-1) unstable; urgency=low

  [ Luca Niccoli ]
  * Initial Debian release. (Closes: #513575)

 -- Luca Niccoli <lultimouomo@gmail.com>  Sat, 03 May 2009 01:47:35 +0200


src.h
/* fswebcam - FireStorm.cx's webcam generator                 */
/*============================================================*/
/* Copyright (C)2005-2011 Philip Heron <phil@sanslogic.co.uk> */
/*                                                            */
/* This program is distributed under the terms of the GNU     */
/* General Public License, version 2. You may use, modify,    */
/* and redistribute it under the terms of this license. A     */
/* copy should be included with this source.                  */

#include <stdint.h>
#include <sys/time.h>

#ifndef INC_SRC_H
#define INC_SRC_H

#define SRC_TYPE_NONE   (0)
#define SRC_TYPE_DEVICE (1 << 0) /* Can capture from a device */
#define SRC_TYPE_FILE   (1 << 1) /* Can capture from a file */

/* When updating the palette list remember to update src_palette[] in src.c */

#define SRC_PAL_ANY     (-1)
#define SRC_PAL_PNG     (0)
#define SRC_PAL_JPEG    (1)
#define SRC_PAL_MJPEG   (2)
#define SRC_PAL_S561    (3)
#define SRC_PAL_RGB32   (4)
#define SRC_PAL_BGR32   (5)
#define SRC_PAL_RGB24   (6)
#define SRC_PAL_BGR24   (7)
#define SRC_PAL_YUYV    (8)
#define SRC_PAL_UYVY    (9)
#define SRC_PAL_YUV420P (10)
#define SRC_PAL_NV12MB  (11)
#define SRC_PAL_BAYER   (12)
#define SRC_PAL_SGBRG8  (13)
#define SRC_PAL_SGRBG8  (14)
#define SRC_PAL_RGB565  (15)
#define SRC_PAL_RGB555  (16)
#define SRC_PAL_Y16     (17)
#define SRC_PAL_GREY    (18)
#define SRC_PAL_NV12    (19)

#define SRC_LIST_INPUTS     (1 << 1)
#define SRC_LIST_TUNERS     (1 << 2)
#define SRC_LIST_FORMATS    (1 << 3)
#define SRC_LIST_CONTROLS   (1 << 4)
#define SRC_LIST_FRAMESIZES (1 << 5)
#define SRC_LIST_FRAMERATES (1 << 6)

/* The SCALE macro converts a value (sv) from one range (sf -> sr)
   to another (df -> dr). */
#define SCALE(df, dr, sf, sr, sv) (((sv - sf) * (dr - df) / (sr - sf)) + df)

typedef struct {
        char    *name;
} src_palette_t;

extern src_palette_t src_palette[];

typedef struct {
	char *name;
	char *value;
} src_option_t;

typedef struct {
	
	/* Source Options */
	char *source;
	uint8_t type;
	
	void *state;
	
	/* Last captured image */
	uint32_t length;
	void *img;
	
	/* Input Options */
	char    *input;
	uint8_t  tuner;
	uint32_t frequency;
	uint32_t delay;
	uint32_t timeout;
	char     use_read;
	
	/* List Options */
	int exposure;
	int hflip;
	int vflip;
	int captured_fps;
	unsigned char src_palette[8];
	
	/* List Options */
	uint8_t list;
	
	/* Image Options */
	int palette;
	uint32_t width;
	uint32_t height;
	uint32_t fps;
	
	src_option_t **option;
	
	/* For calculating capture FPS */
	uint32_t captured_frames;
	struct timeval tv_first;
	struct timeval tv_last;
	
} src_t;

typedef struct {
	
	char *name;
	
	uint8_t flags;
	
	int (*open)(src_t *);
	int (*close)(src_t *);
	int (*grab)(src_t *);
	
} src_mod_t;

extern int src_open(src_t *src, char *source);
extern int src_close(src_t *src);
extern int src_grab(src_t *src);

extern int src_set_option(src_option_t ***options, char *name, char *value);
extern int src_get_option_by_number(src_option_t **opt, int number, char **name, char **value);
extern int src_get_option_by_name(src_option_t **opt, char *name, char **value);
extern int src_free_options(src_option_t ***options);

#endif

